### PR TITLE
Add extra request to fix corner case on github conditional requests

### PR DESCRIPTION
**Pull request by andreugallofre**
### Context

If we request (for example) the pending invitations for a given team /organizations/{org_id}/team/{team_id}/invitations with 30 as the number of items per page, and the number of pending invitations is the same as the number of items per page we get the response with a 200 and the corresponding Etag to handle the cache in our side.

If we invite another person to the team (making that 31 pending invitations) and we do the same request again, with the header If-None-Match with the Etag that we received before, the response is a 304 and the request does not include any pagination headers that are supposed to be there as we now have 1 more page of content.

### CODE REVIEW

File: ghmirror/core/mirror_requests.py
Comment (apahim) [inline ghmirror/core/mirror_requests.py:40]: we already parse the url params at the flask app level, maybe we should just pass them through as a separate argument to the `conditional_requests`:

https://github.com/app-sre/github-mirror/blob/master/ghmirror/app/__init__.py#L87-L91

File: ghmirror/core/mirror_requests.py
Comment (apahim) [inline ghmirror/core/mirror_requests.py:129]: we do not cache ANY response... it has to be a 200 and it has to contain the conditional requests headers so we cache it:

https://github.com/app-sre/github-mirror/blob/master/ghmirror/core/mirror_requests.py#L108-L109

Maybe we should de-duplicate that routine?

File: ghmirror/core/mirror_requests.py
Comment (apahim) [inline ghmirror/core/mirror_requests.py:119]: `if not cached_response.links` instead of `cached_response.links == {}`

File: ghmirror/core/mirror_requests.py
Comment (apahim) [inline ghmirror/core/mirror_requests.py:118]: can we move the cached_response content parsing logic into the `MirrorResponse` class?

**Comment by apahim** [comment]
Awesome workaround! I'd love to see that issue reported to github, not sure what's the channel for that.

**Review by apahim — CHANGES_REQUESTED** [review]
A few minor tweaks, but I agree the mechanism is the best we can do at the moment.

**Comment by apahim** [comment]
Please ping me when you need a second review round :)

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:119]: `cached_response` is not a MirrorResponse, is a "normal" request response, do you mean cast the response to be a MirrorResponse and then implement the logic there?

**Comment by andreugallofre** [comment]
> Awesome workaround! I'd love to see that issue reported to github, not sure what's the channel for that.

I opened a ticket to GitHub support from my account, the number is #1252633 as soon as I have an answer I'll update with more information

File: ghmirror/core/mirror_requests.py
Comment (apahim) [reply ghmirror/core/mirror_requests.py:119]: You are right, I got confused about it. Forget I said anything here :)

**Comment by andreugallofre** [comment]
@apahim Ready for review :)

File: ghmirror/core/mirror_requests.py
Comment (apahim) [inline ghmirror/core/mirror_requests.py:123]: now some params are coming in the `url`, some more are injected with the `params`. A bit hard to keep track.. should w make `url` param-less and only use the `params` arg?

File: tests/functional/test_api.py
Comment (apahim) [inline tests/functional/test_api.py:15]: `links={}` is unsafe (we should probably lint the tests, pylint would tell you so)

File: tests/functional/test_api.py
Comment (apahim) [inline tests/functional/test_api.py:25]: `if self.json_content is not None:`

File: tests/functional/test_api.py
Comment (apahim) [inline tests/functional/test_api.py:461]: :heart: tks for the tests, coverage was kept ay 99% :heart:

File: tests/functional/test_api.py
Comment (andreugallofre) [reply tests/functional/test_api.py]: Solved

File: tests/functional/test_api.py
Comment (apahim) [reply tests/functional/test_api.py]: Sure, sounds like a plan!

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:123]: Not sure how we want to do it. We can strip the URL removing the parameters that come from the client and then add the parameters in the `parameters` dict that's used in the github-mirror request.

Comment (apahim) [reply ghmirror/core/mirror_requests.py:123]: Yeah, I'd not reconstruct the URL in the flask app anymore... Just pass it as param-less url and always use the params argument when we do requests to GitHub. But only if that's not too convoluted.

Comment (apahim) [reply ghmirror/core/mirror_requests.py:123]: Up to you, just a suggestion

**Comment by andreugallofre** [comment]
@apahim All changes done

File: ghmirror/core/mirror_requests.py
Comment (andreugallofre) [reply ghmirror/core/mirror_requests.py:123]: I didn't saw that we were rebuilding the URL, make sense and leave all the parameters in only one place

File: ghmirror/app/__init__.py
Comment (apahim) [inline ghmirror/app/__init__.py:92]: Thank you!

File: ghmirror/core/mirror_requests.py
Comment (apahim) [inline ghmirror/core/mirror_requests.py:41]: Let's use a variable and avoid the double lookup

File: ghmirror/core/mirror_requests.py
Comment (apahim) [inline ghmirror/core/mirror_requests.py:77]: For correctness: `parameters = url_params.to_dict() if url_params is not None else {}`

**Review by apahim — APPROVED** [review]
Looks great, thank you for this! There are two small nitpicks, totally fine if you don't solve them. Feel free to merge either way.

**Pull request merged**
